src/mbpfan.c

```diff
@@ -620,6 +650,13 @@ void mbpfan()
         old_temp = new_temp;
         new_temp = get_temp(sensors);
 
+	if (new_temp <= high_temp) /* maximize turbo below high_temp */
+	    intel_pstate_adjust(intel_pstate, +100);
+	else if (new_temp >= max_temp) /* throttle down to keep the temp in control */
+	    intel_pstate_adjust(intel_pstate, -4);
+	else if ((new_temp - old_temp) <= -3) /* core is cooling down, increase turbo */
+	    intel_pstate_adjust(intel_pstate, +1);
```

Comment:
I tested with `stress --cpu 4` on a MBA 2014 and found that the pstate gradually decreased under load but after removing the load it stayed in one state for a while then jumped by a lot:

```
Adjusting intel_pstate: val: 40, step: -4
Old Temp: 85 New Temp: 86 Fan: Exhaust Speed: 6500
Sleeping for 1 seconds
Old Temp: 86 New Temp: 84 Fan: Exhaust Speed: 6443
Sleeping for 1 seconds
Old Temp: 84 New Temp: 83 Fan: Exhaust Speed: 6386
Sleeping for 1 seconds
Old Temp: 83 New Temp: 81 Fan: Exhaust Speed: 6215
Sleeping for 1 seconds
Old Temp: 81 New Temp: 81 Fan: Exhaust Speed: 6215
Sleeping for 1 seconds
Old Temp: 81 New Temp: 80 Fan: Exhaust Speed: 6101
Sleeping for 1 seconds
Old Temp: 80 New Temp: 80 Fan: Exhaust Speed: 6101
Sleeping for 1 seconds
Old Temp: 80 New Temp: 79 Fan: Exhaust Speed: 5968
Sleeping for 1 seconds
Old Temp: 79 New Temp: 79 Fan: Exhaust Speed: 5968
Sleeping for 1 seconds
Old Temp: 79 New Temp: 78 Fan: Exhaust Speed: 5816
Sleeping for 1 seconds
Old Temp: 78 New Temp: 78 Fan: Exhaust Speed: 5816
Sleeping for 1 seconds
Old Temp: 78 New Temp: 78 Fan: Exhaust Speed: 5816
Sleeping for 1 seconds
Old Temp: 78 New Temp: 77 Fan: Exhaust Speed: 5645
Sleeping for 1 seconds
Old Temp: 77 New Temp: 77 Fan: Exhaust Speed: 5645
Sleeping for 1 seconds
Old Temp: 77 New Temp: 77 Fan: Exhaust Speed: 5645
Sleeping for 1 seconds
Old Temp: 77 New Temp: 76 Fan: Exhaust Speed: 5455
Sleeping for 1 seconds
Old Temp: 76 New Temp: 75 Fan: Exhaust Speed: 5246
Sleeping for 1 seconds
Old Temp: 75 New Temp: 77 Fan: Exhaust Speed: 5246
Sleeping for 1 seconds
Old Temp: 77 New Temp: 77 Fan: Exhaust Speed: 5246
Sleeping for 1 seconds
Old Temp: 77 New Temp: 76 Fan: Exhaust Speed: 5246
Sleeping for 1 seconds
Adjusting intel_pstate: val: 36, step: 1
Old Temp: 76 New Temp: 73 Fan: Exhaust Speed: 4771
Sleeping for 1 seconds
Old Temp: 73 New Temp: 73 Fan: Exhaust Speed: 4771
Sleeping for 1 seconds
Old Temp: 73 New Temp: 72 Fan: Exhaust Speed: 4505
Sleeping for 1 seconds
Old Temp: 72 New Temp: 72 Fan: Exhaust Speed: 4505
Sleeping for 1 seconds
Old Temp: 72 New Temp: 71 Fan: Exhaust Speed: 4220
Sleeping for 1 seconds
Old Temp: 71 New Temp: 70 Fan: Exhaust Speed: 3916
Sleeping for 1 seconds
Old Temp: 70 New Temp: 69 Fan: Exhaust Speed: 3593
Sleeping for 1 seconds
Old Temp: 69 New Temp: 69 Fan: Exhaust Speed: 3593
Sleeping for 1 seconds
Old Temp: 69 New Temp: 69 Fan: Exhaust Speed: 3593
Sleeping for 1 seconds
Old Temp: 69 New Temp: 69 Fan: Exhaust Speed: 3593
Sleeping for 1 seconds
Old Temp: 69 New Temp: 69 Fan: Exhaust Speed: 3593
Sleeping for 1 seconds
Old Temp: 69 New Temp: 69 Fan: Exhaust Speed: 3593
Sleeping for 1 seconds
Old Temp: 69 New Temp: 69 Fan: Exhaust Speed: 3593
Sleeping for 1 seconds
Old Temp: 69 New Temp: 68 Fan: Exhaust Speed: 3251
Sleeping for 1 seconds
Old Temp: 68 New Temp: 68 Fan: Exhaust Speed: 3251
Sleeping for 1 seconds
Old Temp: 68 New Temp: 67 Fan: Exhaust Speed: 2890
Sleeping for 1 seconds
Old Temp: 67 New Temp: 67 Fan: Exhaust Speed: 2890
Sleeping for 1 seconds
Old Temp: 67 New Temp: 68 Fan: Exhaust Speed: 2890
Sleeping for 1 seconds
Old Temp: 68 New Temp: 67 Fan: Exhaust Speed: 2890
Sleeping for 1 seconds
Old Temp: 67 New Temp: 67 Fan: Exhaust Speed: 2890
Sleeping for 1 seconds
Adjusting intel_pstate: val: 37, step: 100
Old Temp: 67 New Temp: 66 Fan: Exhaust Speed: 2510
```

Reply:
I'm guessing you use the default value of `high_temp` in `mbpfan.conf`. This was a deliberate design decision by me to maximize the performance when the temperature is below `high_temp`. Note that setting the pstate doesn't directly change the CPU frequency. An SMT core would opportunistically engage turbo on its own under certain type of loads and then disengage. If a core is stalling on a cache miss for one of its threads, it wouldn't engage turbo for example. If you schedule two compute-bound jobs on the same physical SMT core and the IPC is high, then the core can bet that increasing the clock would provide even more IPC and engage turbo.

Let's take your log for example. The turbo % goes from 40 to 36 at the beginning and the temperature starts falling. At this point, it might be falling because we capped the maximum turbo frequency, or it could be because the load has decreased and turbo isn't kicking in like it used to. There is also the thermal lag, which is apparent in the graph I posted above. Notice how in the middle sections of the graph, every time turbo pct is reduced, it's reflected in the temperature only a few seconds after.

After seeing a temperature drop of 3C (from 76C to 73C), the pstate controller assumes that it's safe to increase the turbo pct by 1% up to 37%. Clearly, if I increase turbo for 1% for every 3C drop in the temperature, I'll never get back to 100%. At low temperatures. you **do** want the turbo to be at 100%. The fans will be enough to handle short, bursty turbo kick-ins that improve performance. So below `high_temp`, I restore it back to 100%. Note again that, restoring it to 100% doesn't affect the CPU frequency under no load, which you might have observed yourself too. Notice how in the graph I posted above, at the end the turbo pct is set to 100% yet the temperature keeps dropping.

Why decrease by 4% at a time? If you look at the beginning of the graph, you'll see that in a manner of seconds the temperature jumps from ~42C to ~77C. With a polling frequency of 1 second and a turbo reduction of 1%, we can never respond to that type of increase in a timely manner. So we decrease aggressively, increase opportunistically. Why increase by 1% only after a 3C drop in temperature? I originally coded with 1C and the change was very jittery. 3C allows for smoother transitions.

The relationship between the turbo and the temperature is unlike the fan speed and the temperature. With the fan, you want higher fan speed at high temperatures, and lower fan speed at lower temperatures. With the turbo, because it's an opportunistic performance mechanism, you want it at maximum at low temperatures **until** the fans are not enough to contain the increase temperature. My mbpro will settle at 79C with a turbo pct of 80% under hours-long compilation workloads. However, for web browsing and things like that, it will have no issues running at 100% turbo.

Sorry this has been a relatively long response to a comment. It's just more intricate than fan control. I'm more than open to suggestions of course.

Reply:
I tested with the latest PR doing a `dnf update`.  I observed similar results and I still don't understand the intended behavior.  In my mind, if the CPU is throttled by a lower pstate, the fan should run as fast as possible to allow the pstate to increase as soon as possible.  I am testing on a MBA 2014 with default config -- maybe we observe different behavior based on hardware?

Reply:
@gktrk Do we have a path forward here?  I still don't understand the intended behavior.